lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp

```diff
@@ -326,6 +372,34 @@ void BlackBoxReaderPass::setOutputFile(VerbatimOp op, StringAttr fileNameAttr) {
     fileListFiles.push_back(fileName);
 }
 
+/// Return true if module is in the DUT hierarchy.
+bool BlackBoxReaderPass::isDut(Operation *module) {
+  // Check if result already cached.
+  auto iter = dutModuleMap.find(module);
+  if (iter != dutModuleMap.end())
+    return iter->getSecond();
+  const StringRef dutAnno = "sifive.enterprise.firrtl.MarkDUTAnnotation";
+  AnnotationSet annos(module);
+  // Any module with the dutAnno, is the DUT.
+  if (annos.hasAnnotation(dutAnno)) {
+    dutModuleMap[module] = true;
+    return true;
+  }
+  auto *node = instanceGraph->lookup(module);
+  if (node->noUses()) {
+    dutModuleMap[module] = false;
+    return false;
+  }
+  // Checking only one of the instances should be enough. Either all the
+  // instances are within DUT or not.
+  auto inst = (*node->usesBegin())->getInstance();
+  // Recursively check the parents.
+  auto dut = isDut(inst->getParentOfType<FModuleOp>());
+  // Cache the result.
+  dutModuleMap[module] = dut;
+  return dut;
```

Comment:
This is likely a dangerous assumption as it would make the result of `isDut` hierarchy dependent? Given the caching going on here, I expect it is fine to do the lookup for all instances and then default to putting it in the dut directory if any instance is in the dut. It would be good to know what the SFC policy is here.

Reply:
Yes, completely agree. Fixed this.